File: richos/tools/richos-service/companion-windows/tests/RichOSCompanionCore.Tests/SessionWriterTests.cs
using System;
using System.IO;
using System.Text.Json;
using RichOSCompanionCore;
using Xunit;

namespace RichOSCompanionCore.Tests
{
    /// <summary>
    /// End-to-end (no audio hardware) proof that <see cref="SessionWriter"/> produces the frozen
    /// contract directory the P1 pipeline consumes: session.json at START (never-silent), a decodable
    /// 2-channel WAV with the correct L/R mapping, health.ndjson, and a populated closed record. Peer
    /// of the macOS companion's <c>SessionWriterTests.swift</c>.
    /// </summary>
    public class SessionWriterTests
    {
        private static SessionContract.Params Params(long startedAt) => new SessionContract.Params(
            sessionId: SessionContract.SessionDirName(startedAt, "system", "call"), startedAt: startedAt,
            sampleRate: 48_000, chunkMs: 3000, micEnabled: true, captureTarget: "system",
            method: SessionContract.MethodSystemLoopback, platformId: "system", platformLabel: "Desktop call",
            platformSlug: "call", companionVersion: "0.1.0-p3", processHint: null);

        [Fact]
        public void StartWritesOpenSessionJsonBeforeAnyAudio()
        {
            string zone = NewZone();
            try
            {
                long t = 1_700_000_000_000;
                // `using`: an open-but-never-closed session must still release its WAV + health handles,
                // otherwise Windows refuses the temp-dir delete (POSIX unlink-open masked this leak).
                using var w = new SessionWriter(zone, Params(t), () => t);
                w.Start();
                string sj = Path.Combine(w.Dir, "session.json");
                Assert.True(File.Exists(sj)); // never-silent: on disk from second zero
                var o = JsonDocument.Parse(File.ReadAllText(sj)).RootElement;
                Assert.Equal("open", o.GetProperty("status").GetString());
                Assert.Equal(0, o.GetProperty("audio").GetProperty("parts").GetArrayLength());
            }
            finally { Directory.Delete(zone, true); }
        }

        [Fact]
        public void FullSessionWritesContractDirWithCorrectChannels()
        {
            string zone = NewZone();
            try
            {
                long virtualNow = 1_700_000_000_000;
                var p = Params(virtualNow);
                using var w = new SessionWriter(zone, p, () => virtualNow);
                w.Start();

                // 1 second of mic=+0.5 (LEFT) and system=-0.5 (RIGHT).
                var mic = Fill(48_000, 0.5f);
                var sys = Fill(48_000, -0.5f);
                w.PushMic(mic);
                w.PushSystem(sys);
                w.Pump();
                w.WriteHealthTick(0.5, 0.5, 0.5, 0.5, tapRunning: true, micRunning: true);
                virtualNow += 1000;
                w.Close();

                // session.json closed + populated
                var o = JsonDocument.Parse(File.ReadAllText(Path.Combine(w.Dir, "session.json"))).RootElement;
                Assert.Equal("closed", o.GetProperty("status").GetString());
                Assert.True(o.GetProperty("audio").GetProperty("bytesTotal").GetInt32() > 0);

                // health.ndjson present with at least one row
                string[] health = File.ReadAllLines(Path.Combine(w.Dir, "health.ndjson"));
                Assert.True(health.Length >= 1);

                // the WAV decodes and preserves LEFT=mic(+0.5), RIGHT=system(-0.5)
                var pcm = WavReader.Read(Path.Combine(w.Dir, "audio-part-00.wav"));
                Assert.Equal(2, pcm.Channels);
                Assert.True(pcm.ChannelsData[0][0] > 0.49f && pcm.ChannelsData[0][0] < 0.51f);   // LEFT/mic
                Assert.True(pcm.ChannelsData[1][0] < -0.49f && pcm.ChannelsData[1][0] > -0.51f); // RIGHT/system
            }
            finally { Directory.Delete(zone, true); }
        }

        private static float[] Fill(int n, float v)
        {
            var a = new float[n];
            for (int i = 0; i < n; i++) a[i] = v;
            return a;
        }

        private static string NewZone()
        {
            string zone = Path.Combine(Path.GetTempPath(), "richos-zone-" + Guid.NewGuid().ToString("N"));
            Directory.CreateDirectory(zone);
            return zone;
        }
    }
}

File: richos/tools/richos-service/companion-macos/Tests/RichOSCompanionCoreTests/SessionWriterTests.swift
import XCTest
@testable import RichOSCompanionCore

/// End-to-end (pure-Swift) proof that the SessionWriter produces a valid contract directory: the WAV
/// is a decodable 2-channel 16-bit file, session.json flips open->closed with populated audio, and
/// health.ndjson carries one row per tick. (The REAL P1 transcription handoff is proven separately
/// by the on-Mac `ingest` + `node bin/richos-service.js run` E2E — see README.)
final class SessionWriterTests: XCTestCase {

    func testWritesValidContractDirectory() throws {
        let tmp = NSTemporaryDirectory() + "richos-companion-test-\(UUID().uuidString)"
        defer { try? FileManager.default.removeItem(atPath: tmp) }

        let startedAt: Int64 = 1_700_000_000_000
        var clock = startedAt
        let params = SessionContract.Params(
            sessionId: SessionContract.sessionDirName(startedAt: startedAt, platformId: "system", slug: "call"),
            startedAt: startedAt, sampleRate: 16_000, chunkMs: 3000, micEnabled: true,
            captureTarget: "system", platformId: "system", platformLabel: "Desktop call",
            platformSlug: "call", companionVersion: "test", processHint: nil)

        let writer = SessionWriter(rootZone: tmp, params: params, clock: { clock })
        try writer.start()

        let dir = (tmp as NSString).appendingPathComponent(params.sessionId)
        // session.json exists at START with status open.
        let openData = try Data(contentsOf: URL(fileURLWithPath: (dir as NSString).appendingPathComponent("session.json")))
        let openObj = try JSONSerialization.jsonObject(with: openData) as! [String: Any]
        XCTAssertEqual(openObj["status"] as? String, "open")

        // Feed 3 seconds of aligned tone; mic=LEFT, system=RIGHT.
        for _ in 0..<3 {
            let mic = (0..<16_000).map { sinf(Float($0) * 0.05) * 0.5 }
            let sys = (0..<16_000).map { sinf(Float($0) * 0.10) * 0.5 }
            writer.pushMic(mic); writer.pushSystem(sys)
            try writer.pump()
            try writer.writeHealthTick(micPeak: 0.5, micRms: 0.35, sysPeak: 0.5, sysRms: 0.35,
                                       tapRunning: true, micRunning: true)
            clock += 1000
        }
        try writer.close()

        // Closed session.json: audio populated.
        let closedData = try Data(contentsOf: URL(fileURLWithPath: (dir as NSString).appendingPathComponent("session.json")))
        let closedObj = try JSONSerialization.jsonObject(with: closedData) as! [String: Any]
        XCTAssertEqual(closedObj["status"] as? String, "closed")
        let audio = closedObj["audio"] as? [String: Any]
        XCTAssertGreaterThan(audio?["bytesTotal"] as? Int ?? 0, 3 * 16_000 * 4 - 100) // ~3s stereo 16-bit

        // The WAV part decodes as 2-channel 16-bit at 16 kHz with ~3s of data.
        let wavPath = (dir as NSString).appendingPathComponent("audio-part-00.wav")
        let pcm = try WavReader.read(path: wavPath)
        XCTAssertEqual(pcm.channels, 2)
        XCTAssertEqual(pcm.sampleRate, 16_000)
        XCTAssertEqual(pcm.channelsData.count, 2)
        XCTAssertGreaterThan(pcm.channelsData[0].count, 16_000 * 2) // > 2s of frames

        // health.ndjson has one row per tick.
        let health = try String(contentsOf: URL(fileURLWithPath: (dir as NSString).appendingPathComponent("health.ndjson")), encoding: .utf8)
        let rows = health.split(separator: "\n").filter { !$0.isEmpty }
        XCTAssertEqual(rows.count, 3)
    }

    func testWavWriterHeaderIsValidAfterFlush() throws {
        let path = NSTemporaryDirectory() + "richos-wav-\(UUID().uuidString).wav"
        defer { try? FileManager.default.removeItem(atPath: path) }
        let w = try WavWriter(path: path, sampleRate: 48_000, channels: 2, bitsPerSample: 16)
        try w.append(ChannelMixer.int16LEBytes([1, 2, 3, 4])) // 2 frames
        try w.close()
        let pcm = try WavReader.read(path: path)
        XCTAssertEqual(pcm.sampleRate, 48_000)
        XCTAssertEqual(pcm.channels, 2)
        XCTAssertEqual(pcm.channelsData[0].count, 2)
    }
}
